src/core/surface/call.c

```diff
@@ -425,7 +425,13 @@ static grpc_cq_completion *allocate_completion(grpc_call *call) {
     if (call->allocated_completions & (1u << i)) {
       continue;
     }
-    call->allocated_completions |= (gpr_uint8)(1u << i);
+    /* NB: the following integer arithmetic operation needs to be in its
+     * expanded form due to the "integral promotion" performed (see section
+     * 3.2.1.1 of the C89 draft standard), which in this case upcasts the result
+     * of the bitwise OR to "unsigned". A cast to the smaller container type is
```

Comment:
But after it gets typecast to (gpr_uint8), it ends up as a unsigned char.
That doesn't trigger any warning because it is an explicit conversion. The
conversion warning is happening because it then promotes to int, gets
or'ed, and then gets assigned to unsigned char.

On Tue, Oct 6, 2015 at 10:21 AM David G. Quintas <notifications@github.com>
wrote:

> In src/core/surface/call.c
> <https://github.com/grpc/grpc/pull/3615#discussion_r41294148>:
>
> > @@ -425,7 +425,13 @@ static grpc_cq_completion *allocate_completion(grpc_call *call) {
> >      if (call->allocated_completions & (1u << i)) {
> >        continue;
> >      }
> > -    call->allocated_completions |= (gpr_uint8)(1u << i);
> > +    /* NB: the following integer arithmetic operation needs to be in its
> > +     * expanded form due to the "integral promotion" performed (see section
> > +     * 3.2.1.1 of the C89 draft standard), which in this case upcasts the result
> > +     * of the bitwise OR to "unsigned". A cast to the smaller container type is
>
> We are talking about different things. I think you are talking about
> call->allocated_completions being promoted to int prior to the bitwise or.
> I'm talking about the whole bitwise or expression in parenthesis, whose
> result is an unsigned because of of its operands is unsigned, namely (1u <<
> i).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc/pull/3615/files#r41294148>.
>